src/librustc_resolve/lib.rs

```diff
@@ -2390,23 +2399,36 @@ impl<'a> Resolver<'a> {
                     binding.res().descr(),
                     ident.name,
                 );
-                // FIXME: use the ctor's `def_id` to check wether any of the fields is not visible
-                match binding.kind {
+
+                let def_id = match binding.kind {
                     NameBindingKind::Res(Res::Def(DefKind::Ctor(
                         CtorOf::Struct,
                         CtorKind::Fn,
-                    ), _def_id), _) => {
+                    ), def_id), _) => {
                         err.note("a tuple struct constructor is private if any of its fields \
                                   is private");
+                        Some(def_id)
                     }
                     NameBindingKind::Res(Res::Def(DefKind::Ctor(
                         CtorOf::Variant,
                         CtorKind::Fn,
-                    ), _def_id), _) => {
+                    ), def_id), _) => {
                         err.note("a tuple variant constructor is private if any of its fields \
                                   is private");
+                        Some(def_id)
+                    }
+                    _ => None,
+                };
+
+                if let Some(def_id) = def_id {
+                    if let Some(fields) = self.field_vis.get(&def_id) {
+                        for field in fields {
+                            match (field.vis, field.span) {
+                                (ty::Visibility::Public, _) => {},
+                                (_, span) => { err.span_label(span, "field is private"); },
```

Comment:
This check is not correct in general, the correct thing would be to check whether the field is private from the point of use.
However, just pointing to all fields would be much simpler and good enough.

Reply:
Ok, so to pointing to all fields, I'm guessing we need to store `sp` while in `build_reduced_graph_for_item`, and later in `report_errors` use that inside the preceding two `NameBindingKind::Res` matchers while removing the incorrect check.